genie-common-internal/src/main/java/com/netflix/genie/common/internal/dto/v4/JobRequestMetadata.java

```diff
@@ -45,19 +45,32 @@
     private final long totalSizeOfAttachments;
 
     /**
-     * Constructor.
+     * Constructor. One of {@literal apiClientMetadata} or {@literal agentClientMetadata} is required. Both cannot be
+     * present.
      *
      * @param apiClientMetadata      The metadata about the client if this request was received via API
      * @param agentClientMetadata    The metadata about the client if this request was received from the Agent
      * @param numAttachments         The number of attachments that came with this job request
      * @param totalSizeOfAttachments The total size of the attachments that came with this job request
+     * @throws IllegalArgumentException If both {@literal apiClientMetadata} and {@literal agentClientMetadata} are
+     *                                  missing or both are present
      */
     public JobRequestMetadata(
         @Nullable final ApiClientMetadata apiClientMetadata,
         @Nullable final AgentClientMetadata agentClientMetadata,
         final int numAttachments,
         final long totalSizeOfAttachments
-    ) {
+    ) throws IllegalArgumentException {
+        if (apiClientMetadata == null && agentClientMetadata == null) {
+            throw new IllegalArgumentException(
+                "Both apiClientMetadata and agentClientMetadata are missing. One is required."
+            );
+        }
+        if (apiClientMetadata != null && agentClientMetadata != null) {
```

Comment:
I'm pretty sure.

This is within `JobRequestMetadata` and is only used to capture/return data about the initial request not subsequent execution metadata. There's no way (that I can think of right now) that a Job Request could come from both the API and some other mechanism.

Doing this allows us to not have to pass in an `isApi` `boolean` to this object as well as we can derive it based on which one of the metadata objects is passed in. I've considered other options like Template class on one of these metadata types but it would be a pain.

If you can think of a good reason we'd need both at initial job submission time we can remove this commit.

I looked at the job service and perhaps you're mixing this with `claimJob` which sends the `AgentMetadata` again and we persist it within `JobPersistenceService.claimJob`